File: DSA/Manav-Rajput/Day73_question1.java
class Solution {
    int leftSubtreeSize = 0;
    int rightSubtreeSize = 0;

    public boolean btreeGameWinningMove(TreeNode root, int n, int x) {
        // Find the sizes of the left and right subtrees of node x
        countSubtreeNodes(root, x);

        // The remaining nodes excluding the subtree rooted at x
        int rest = n - (leftSubtreeSize + rightSubtreeSize + 1);

        // Check if the second player can control more than half the nodes
        if (leftSubtreeSize > n / 2 || rightSubtreeSize > n / 2 || rest > n / 2) {
            return true;
        }

        return false;
    }

    private int countSubtreeNodes(TreeNode node, int x) {
        if (node == null) return 0;

        int left = countSubtreeNodes(node.left, x);
        int right = countSubtreeNodes(node.right, x);

        if (node.val == x) {
            leftSubtreeSize = left;
            rightSubtreeSize = right;
        }

        return left + right + 1;
    }
}

File: DSA/Arshdeep-Kaur/Day 73/Day73_Question1.cpp
class Solution {
public:
    int count(TreeNode* root, int x, int &lc, int &rc) {
        if (root == NULL) {
            return 0;
        }

        int left = count(root->left, x, lc, rc);
        int right = count(root->right, x, lc, rc);
        if(root->val == x) {
            lc = left;
            rc = right;
        }

        return left+right+1;
    }

    bool btreeGameWinningMove(TreeNode* root, int n, int x) {
        int lc = 0, rc = 0;
        count(root, x, lc, rc);

        int nnpex = n - (lc+rc+1);
        int maxi = max(nnpex, max(lc, rc));

        return maxi>(n/2);
    }
};
